Add Accessor for views and buffers (only feature) (#1433)
alpaka-group/alpaka: docs/source/basic/library.rst

Additionally the memory buffer objects know their extents, their pitches as well as the device they reside on.
This allows buffers that possibly reside on different devices with different pitches to be copied only by providing the buffer objects as well as the extents of the region to copy (``alpaka::memcpy(bufDevA, bufDevB, copyExtents``).

Accessors
`````````

An accessor is an interface to access the data stored by a buffer or, more generally, a view.
Accessors take care of the multidimensionality of their underlying buffers when indexed.
It is created via one of the following calls:

.. code-block:: cpp

   auto accessor = alpaka::experimental::access(buffer);      // read/write
   auto accessor = alpaka::experimental::readAccess(buffer);  // read
   auto accessor = alpaka::experimental::writeAccess(buffer); // write

Accessors have many template parameter and users are adviced to use the ``BufferAccessor`` convenience alias to get the type of an accessor for a buffer of a given accelerator.
Example kernel with 3D accessor:

.. code-block:: c++

   struct Kernel {
      template<typename Acc>

>       ALPAKA_FN_ACC void operator()(Acc const & acc, alpaka::experimental::BufferAccessor<Acc, float, 3, alpaka::experimental::ReadAccess> data) const {

Reviewer: I think, this should be `alpaka::experimental::writeAccess`

If yes, the `alpaka::exec` needs also be modified.

Author: Good catch! Thx!

         ...

>          data(z, y, x) = 42.0f;

Reviewer: In my opinion, the example should be extend by the use of `extents`. 

```
for(auto z = 0; z < data.extents[0];  ++z)
   for(auto y = 0; y < data.extents[1]; ++y)
     for(auto x = 0; x < data.extents[2];  ++x)
       data(z, y, x) = 42.0f;
```

By the way, does `accessors` respect the pitching? You should document this also in small sentence.

Author: Good point! Yes the accessor does respect the pitch, that is the whole point of it :D

Additionally the memory buffer objects know their extents, their pitches as well as the device they reside on.
This allows buffers that possibly reside on different devices with different pitches to be copied only by providing the buffer objects as well as the extents of the region to copy (``alpaka::memcpy(bufDevA, bufDevB, copyExtents``).

Accessors
`````````

An accessor is an interface to access the data stored by a buffer or, more generally, a view.
Accessors take care of the multidimensionality of their underlying buffers when indexed, including pitched allocations.
It is created via one of the following calls:

.. code-block:: cpp

>    auto accessor = alpaka::experimental::access(buffer);      // read/write
>    auto accessor = alpaka::experimental::readAccess(buffer);  // read
>    auto accessor = alpaka::experimental::writeAccess(buffer); // write

Reviewer: ```suggestion
   auto accessor = alpaka::experimental::ReadWriteAccess(buffer);      // read/write
   auto accessor = alpaka::experimental::ReadAccess(buffer);  // read
   auto accessor = alpaka::experimental::WriteAccess(buffer); // write
```

Author: I get the suggestion to rename `access` to `readWriteAccess`. But these are functions, so I would like to keep them lowercase.

Reviewer: Sorry, I mixed up the function, where you get the accessor on the host side with the accessor types in an alpaka kernel. Therefore the documentation is correct and my suggested change is wrong.

There is only one small point. I mixed up the host and kernel accessor, because the only difference between the host and the device version of the read and write accessor is the case of the first letter. Should we think about, to make the names more unique or do you think, this is no problem?

Author: I think you mixed up the second point as well. There is no difference between host and device accessors. And there shouldn't be one. You always produce and `Accessor` from an existing view or accessor with `access(...)`.

Reviewer: Yes, I mixed every point up. So this means `access`, `readAccess` and `writeAccess` are factory methods? In this case, I would suggest to replace the `auto` with actual type to document the relationship.

```suggestion
   alpaka::experimental::BufferAccessor<Acc, float, 3, alpaka::experimental::ReadWriteAccess> accessor = 
     alpaka::experimental::access(buffer);
   alpaka::experimental::BufferAccessor<Acc, float, 3, alpaka::experimental::ReadAccess> accessor = 
     alpaka::experimental::readAccess(buffer);  
   alpaka::experimental::BufferAccessor<Acc, float, 3, alpaka::experimental::WriteAccess> accessor = 
     alpaka::experimental::writeAccess(buffer); 
```

Author: I would like to avoid that as well, because `access` only returns a `BufferAccessor` for the current alpaka buffers. It returns something else for views I think (a different instantiation of `Accessor` than the `BufferAccessor` alias). Also SYCL might change the landscape here. Users should definitely use `auto` in there code.

Reviewer: Okay, that's one point. But my motivation for the concrete type instead of `auto` is that I know what I'm getting and I can see in the documentation what the member function and variable of the type are and where I can use them. But this is not possible if the return type is not specified. That's why @sbastrakov suggests to use a "concepts-structs".

> E.g. this functor: https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/include/picongpu/fields/MaxwellSolver/Yee/StencilFunctor.hpp . Btw note that the T_Curl template parameter is also from Curl "concept". It's used here: https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/include/picongpu/fields/MaxwellSolver/FDTD/FDTD.kernel

In our case, maybe we can create a concept like:

```c++
struct ConceptAccessor {
public:
   //! Store dimension and size of each dimension
   Vec<...> extends
};
``` 

Than you can write in the documentation, that the return type has to fulfill the `ConceptAccessor` and point to the doxygen documentation.

And as bonus, if the some guy makes pressure, that we should switch to C++20 (and know, who it is ;-)), we can easily replace the concept-structs with real c++ concepts.

Author: The return type of `access` et al. is a an object which is callable on the number of dimensions and has a member `extends`, which is subscriptable  in the number of dimensions. This should also be obvious by the example below (library.rst:237) so I think an additional concept or concept struct is of little value.

I also intentionally keep the documentation of this feature short for now, because it is experimental and we for sure want to change it some times again. I am almost certain it will change after merging when @j-stephan will rebase SYCL on top of it. So I would like to wait with more user side information until the details are fleshed out.

> Accessors have many template parameter and users are adviced to use the ``BufferAccessor`` convenience alias to get the type of an accessor for a buffer of a given accelerator.

Reviewer 2: Two typos in this line:

```suggestion
Accessors have many template parameters and users are advised to use the ``BufferAccessor`` convenience alias to get the type of an accessor for a buffer of a given accelerator.
```

Not something that needs to be fixed right now but we should keep it in mind for the main accessor PR.

Author: thx!

Example kernel with 3D accessor:

.. code-block:: c++

   struct Kernel {
      template<typename Acc>

>       ALPAKA_FN_ACC void operator()(Acc const & acc, alpaka::experimental::BufferAccessor<Acc, float, 3, alpaka::experimental::WriteAccess> data) const {

Reviewer 2: There should probably be a line break somewhere here.

Author: thx!

         ...
         for(Idx z = 0; z < data.extents[0];  ++z)
             for(Idx y = 0; y < data.extents[1]; ++y)
                 for(Idx x = 0; x < data.extents[2];  ++x)

>                      data(z, y, x) = 42.0f;

Reviewer 3: This doesn't seem like a good example ?
I would assume that one wants different threads to write to different positions inside the buffer, not that all threads write to the whole buffer (without using atomic operations).

Author: For sure it can be improved! But it is good enough to showcase the API in a small amount of LOCs. More elaborate examples can be found once some of the standard alpaka examples are rewritten using accessors, see #1458.
